### ISSUE TITLE
FFJORD

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/DiffEqFlux.jl#47]
```julia
using OrdinaryDiffEq
using Distributions
using Flux, DiffEqFlux, Tracker

# Neural Network
nn = Chain(Dense(1,1,tanh))
p = Tracker.data(DiffEqFlux.destructure(nn))
DiffEqFlux.restructure(nn,p)([1.0])
tspan = Float32.((0.0, 10.0))
function cnf(du,u,p,t)
  z = @view u[1:end-1]
  m = DiffEqFlux.restructure(nn,p)
  du[1:end-1] = m(z)
  fz, back = Tracker.forward(z->m(z),z')
  e = randn(size(z)[1])
  eJ = back(e)[1]
  eJe = Tracker.data((eJ .* e)[1])
  du[end] = -eJe
end

Tracker.collectmemaybe(z)

prob = ODEProblem(cnf,nothing,tspan,nothing)

p = param(Float32[0.0, 0.0]) # Initial Parameter Vector
params = Params([p])

function predict_adjoint(x)
    diffeq_adjoint(p,prob,Tsit5(),u0=[x;false],
                   saveat=0.0:0.1:10.0,
                   sensealg=DiffEqFlux.SensitivityAlg(quad=false,
                                backsolve=true,autojacvec=true))
end

function loss_adjoint(xs)
    pz = Normal(0.0, 1.0)
    preds = [predict_adjoint(x)[:,end] for x in xs]
    z = [pred[1] for pred in preds] # TODO better slicing
    delta_logp = [pred[2] for pred in preds]

    logpz = logpdf.(pz, z)
    logpx = logpz - delta_logp
    loss = -mean(logpx)
end

opt = ADAM(0.1)

raw_data = [Float32[rand(Normal(2.0, 0.1)) for i in 1:100]]
data = Iterators.repeated(raw_data, 100);

Flux.train!(loss_adjoint, params, data, opt)

# check whether it looks standard normal
using Plots

preds = [predict_adjoint(r)[:,end] for r in raw_data[1]];
```

**Comment by ChrisRackauckas** [comment]
```julia
using OrdinaryDiffEq
using Distributions
using Flux, DiffEqFlux, Tracker

# Neural Network
nn = Dense(1,1,tanh)
p = Tracker.data(DiffEqFlux.destructure(nn))
tspan = Float32.((0.0, 10.0))
function ffjord(du,u,p,t)
  z = @view u[1:end-1]
  m = DiffEqFlux.restructure(nn,p)
  fz, back = Tracker.forward(z->m(z),z')
  e = randn(size(z)[1])
  eJ = back(e)[1]
  eJe = (eJ .* e)[1]
  if u isa TrackedArray
      [m(z);-eJe]
  else
      Tracker.data([m(z);-eJe])
  end
end

prob = ODEProblem(ffjord,nothing,tspan,nothing)

p = param(Float32[0.0, 0.0]) # Initial Parameter Vector
params = Params([p])

function predict_adjoint(x)
    diffeq_adjoint(p,prob,Tsit5(),u0=[x;false],
                   saveat=0.0:0.1:10.0,
                   sensealg=DiffEqFlux.SensitivityAlg(quad=false,
                                backsolve=true,autojacvec=true))
end

function loss_adjoint(xs)
    pz = Normal(0.0, 1.0)
    preds = [predict_adjoint(x)[:,end] for x in xs]
    z = [pred[1] for pred in preds] # TODO better slicing
    delta_logp = [pred[2] for pred in preds]

    logpz = logpdf.(pz, z)
    logpx = logpz - delta_logp
    loss = -mean(logpx)
end

opt = ADAM(0.1)

raw_data = [Float32[rand(Normal(2.0, 0.1)) for i in 1:100]]
data = Iterators.repeated(raw_data, 100);

loss_adjoint(raw_data[1])
Flux.train!(loss_adjoint, params, data, opt)
iszero(Tracker.grad(nn[1].W))
loss_adjoint(raw_data[1])
```

works with https://github.com/FluxML/Tracker.jl/pull/24

**Comment by GJBoth** [comment]
I'm working on implementing FFJORD (or at least CNF) too, but I can't get the above code to train, i.e. it runs but my loss doesn't decrease on Julia 1.1.1, Flux 0.8.3, DiffEqFlux 0.6.1.

Secondly, it seems this implementation creates a new random vector each call to ffjord; in the paper they suggest sampling a noise vector once and then using that. This means that the noise vector e should be added to the parameters and passed each call instead of sampled each call.

**Comment by GJBoth** [comment]
Thanks for the quick reply. And other autodiff packages such as Zygote, do they work? Also, by not nesting well do you think it's pointless for me to get an implementation going untill the autodiff is fixed or is it just going to be non-optimal implementation-wise?

**Comment by GJBoth** [comment]
I'm fairly new to Julia (so still finding my way around), so I think adding that option is out of my league for now. If you suspect things would work Zygote, I'll give that a try and I'll keep you updated.

**Comment by jessebett** [comment]
>Secondly, it seems this implementation creates a new random vector each call to ffjord; in the paper they suggest sampling a noise vector once and then using that. This means that the noise vector e should be added to the parameters and passed each call instead of sampled each call.

@GJBoth you're completely correct. This is a mistake in my implementation. The sample should not happen inside the integrand. (Though in expectation these are equivalent, the variance properties will be much worse the way I have it.)

**Comment by ChrisRackauckas** [comment]
With the advance in https://github.com/JuliaDiffEq/DiffEqFlux.jl/issues/46 , we might be able to fix this one up now.

**Comment by ChrisRackauckas** [comment]
An updated working version:

```julia
using OrdinaryDiffEq
using Distributions, DistributionsAD
using Flux, DiffEqFlux, Tracker

# Neural Network
nn = Chain(Dense(1,1,tanh))
θ,re = Flux.destructure(nn)
tspan = Float32.((0.0, 10.0))
function ffjord(u,p,t)
  z = @view u[1:end-1]
  m = re(p)
  fz, back = Zygote.pullback(z->m(z),z')
  e = randn(Float32,size(z)[1])
  eJ = back(e)[1]
  eJe = (eJ .* e)
  vec([m(z);-eJe])
end

prob = ODEProblem(ffjord,nothing,tspan,nothing)

function predict_adjoint(θ,x)
    concrete_solve(prob,Tsit5(),[x;false],θ,
                   saveat=0f0:0.1f0:10f0,
                   sensealg=DiffEqSensitivity.InterpolatingAdjoint(autojacvec=TrackerVJP()))
end

function loss_adjoint(θ,xs)
    pz = Normal(0f0, 1f0)
    preds = [predict_adjoint(θ,x)[:,end] for x in xs]
    z = [pred[1] for pred in preds] # TODO better slicing
    delta_logp = [pred[2] for pred in preds]

    logpz = logpdf.(pz, z)
    logpx = logpz - delta_logp
    loss = -mean(logpx)
end

function cb(θ,l)
    @show l
    false
end

opt = ADAM(0.01)
raw_data = Iterators.cycle(([Float32[rand(Normal(2.0, 0.1)) for i in 1:100]],))
DiffEqFlux.sciml_train(loss_adjoint, θinit, opt, raw_data, cb=cb, maxiters=100)
```

Time to library-itize it.
